homeassistant/components/roon/config_flow.py

```diff
@@ -0,0 +1,102 @@
+"""Config flow for roon integration."""
+import logging
+
+from roon import RoonApi
+import voluptuous as vol
+
+from homeassistant import config_entries, core, exceptions
+from homeassistant.const import CONF_API_KEY, CONF_HOST
+
+from .const import DEFAULT_NAME, DOMAIN, ROON_APPINFO  # pylint: disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema({"host": str})
+
+
+class RoonHub:
+    """Interact with roon during config flow."""
+
+    def __init__(self, host):
+        """Initialize."""
+        self._host = host
+
+    async def authenticate(self) -> bool:
+        """Test if we can authenticate with the host."""
+        try:
+            roonapi = RoonApi(ROON_APPINFO, None, self._host)
+        except Exception as error:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception %s", error)
+            raise CannotConnect
+
+        token = roonapi.token
+        roonapi.stop()
+        return token
+
+
+async def authenticate(hass: core.HomeAssistant, host):
+    """Connect and authenticate home assistant."""
+
+    hub = RoonHub(host)
+    token = await hub.authenticate()
+    if token is None:
+        raise InvalidAuth
+
+    return {CONF_HOST: host, CONF_API_KEY: token}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for roon."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    def __init__(self):
+        """Initialize the Roon flow."""
+        self._host = None
+
+    async def async_step_user(self, user_input=None):
+        """Handle getting host details from the user."""
+
+        errors = {}
+        if user_input is not None:
+            try:
+                self._host = user_input["host"]
+                return await self.async_step_link()
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+        return self.async_show_form(
+            step_id="user", data_schema=DATA_SCHEMA, errors=errors
+        )
+
+    async def async_step_link(self, user_input=None):
+        """Handle linking and authenticting with the roon server."""
+
+        errors = {}
+        if user_input is not None:
+            try:
+                info = await authenticate(self.hass, self._host)
+                return self.async_create_entry(title=DEFAULT_NAME, data=info)
```

Comment:
Can we get some id from the api to identify the device, besides host? Then we could use that as unique id for the config entry.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-ids

The id needs to be both unique and stable.

Reply:
I've had a dig through the api - and sadly looks like Roon just use the hostname and ip!

Reply:
Then we need to check the new ip address against the existing entry data ip addresses so that we don't set up the same ip more than once.

Reply:
Also please move the entry creation outside the try... except block. We can use an `else:` block if needed.

Reply:
@MartinHjelmare Could you give me a pointer on how to get the details of an existing integration. I'm sure it's in the docs - but struggled to find the info.

Reply:
Call `self._async_current_entries` inside the config flow to get all entries for the integration domain. Then we can check entry.data for whatever item and compare that with some value that we want to filter on.

Reply:
Thanks for the pointer, have done as you suggested.